In [190]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

In [191]:
SHG_Booking=pd.read_csv('SHG_Booking_Dataa.csv')

SHG_Booking

Booking ID   Hotel    Booking Date    Arrival Date  Lead Time  Nights  \
0                1  Resort  7/24/2014 0:00   7/1/2015 0:00        342       0   
1                2  Resort  6/24/2013 0:00   7/1/2015 0:00        737       0   
2                3  Resort  6/24/2015 0:00   7/1/2015 0:00          7       1   
3                4  Resort  6/18/2015 0:00   7/1/2015 0:00         13       1   
4                5  Resort  6/17/2015 0:00   7/1/2015 0:00         14       2   
...            ...     ...             ...             ...        ...     ...   
119385      119386    City   8/7/2017 0:00  8/30/2017 0:00         23       7   
119386      119387    City  5/21/2017 0:00  8/31/2017 0:00        102       7   
119387      119388    City  7/28/2017 0:00  8/31/2017 0:00         34       7   
119388      119389    City  5/14/2017 0:00  8/31/2017 0:00        109       7   
119389      119390    City   2/5/2017 0:00  8/29/2017 0:00        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            2                Direct     Transient        Portugal   
1            2                Direct     Transient        Portugal   
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status  Status Update  \
0        No Deposit               0  Check-Out  7/1/2015 0:00   
1        No Deposit               0  Check-Out  7/1/2015 0:00   
2        No Deposit              75  Check-Out  7/2/2015 0:00   
3        No Deposit              75  Check-Out  7/2/2015 0:00   
4        No Deposit              98  Check-Out  7/3/2015 0:00   
...             ...             ...        ...            ...   
119385   No Deposit              96  Check-Out  9/6/2017 0:00   
119386   No Deposit             225  Check-Out  9/7/2017 0:00   
119387   No Deposit             158  Check-Out  9/7/2017 0:00   
119388   No Deposit             104  Check-Out  9/7/2017 0:00   
119389   No Deposit             151  Check-Out  9/7/2017 0:00   

        Cancelled (0/1)  Revenue  Revenue Loss  
0                     0        0             0  
1                     0        0             0  
2                     0       75             0  
3                     0       75             0  
4                     0      196             0  
...                 ...      ...           ...  
119385                0      673             0  
119386                0     1578             0  
119387                0     1104             0  
119388                0      731             0  
119389                0     1361             0  

[119390 rows x 17 columns]

In [192]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   Booking ID            119390 non-null  int64 
 1   Hotel                 119390 non-null  object
 2   Booking Date          119390 non-null  object
 3   Arrival Date          119390 non-null  object
 4   Lead Time             119390 non-null  int64 
 5   Nights                119390 non-null  int64 
 6   Guests                119390 non-null  int64 
 7   Distribution Channel  119390 non-null  object
 8   Customer Type         119390 non-null  object
 9   Country               118902 non-null  object
 10  Deposit Type          119390 non-null  object
 11  Avg Daily Rate        119390 non-null  int64 
 12  Status                119390 non-null  object
 13  Status Update         119390 non-null  object
 14  Cancelled (0/1)       119390 non-null  int64 
 15  Revenue          

After checking the data info, I noticed that the data types for Booking Date, Arrival Date, and Status Update were wrong. So I decided to change them.

In [193]:
SHG_Booking['Booking Date']=pd.to_datetime(SHG_Booking['Booking Date'])
SHG_Booking['Arrival Date']=pd.to_datetime(SHG_Booking['Arrival Date'])
SHG_Booking['Status Update']=pd.to_datetime(SHG_Booking['Status Update'])

Also, I noticed the values, by checking with the Microsoft Excel, that most of the values of the column "Cancelled (0/1)" are between 0 and 1. So, I decided to change the data type to "bool" but there is an hindrance. What is it? I am not sure if all the data input in the column are values between 1 and 0. So, I decided to check and confirm before proceeding. 

In [194]:
SHG_Booking['Cancelled (0/1)'].unique()

array([0, 1], dtype=int64)

The confimation shows a positive sign to change the data type to "bool" and also I want to change the column name to a easier one.

In [195]:
SHG_Booking.rename(columns={'Cancelled (0/1)':'Cancelled'})


Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
0                1  Resort   2014-07-24   2015-07-01        342       0   
1                2  Resort   2013-06-24   2015-07-01        737       0   
2                3  Resort   2015-06-24   2015-07-01          7       1   
3                4  Resort   2015-06-18   2015-07-01         13       1   
4                5  Resort   2015-06-17   2015-07-01         14       2   
...            ...     ...          ...          ...        ...     ...   
119385      119386    City   2017-08-07   2017-08-30         23       7   
119386      119387    City   2017-05-21   2017-08-31        102       7   
119387      119388    City   2017-07-28   2017-08-31         34       7   
119388      119389    City   2017-05-14   2017-08-31        109       7   
119389      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            2                Direct     Transient        Portugal   
1            2                Direct     Transient        Portugal   
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update  Cancelled  \
0        No Deposit               0  Check-Out    2015-07-01          0   
1        No Deposit               0  Check-Out    2015-07-01          0   
2        No Deposit              75  Check-Out    2015-07-02          0   
3        No Deposit              75  Check-Out    2015-07-02          0   
4        No Deposit              98  Check-Out    2015-07-03          0   
...             ...             ...        ...           ...        ...   
119385   No Deposit              96  Check-Out    2017-09-06          0   
119386   No Deposit             225  Check-Out    2017-09-07          0   
119387   No Deposit             158  Check-Out    2017-09-07          0   
119388   No Deposit             104  Check-Out    2017-09-07          0   
119389   No Deposit             151  Check-Out    2017-09-07          0   

        Revenue  Revenue Loss  
0             0             0  
1             0             0  
2            75             0  
3            75             0  
4           196             0  
...         ...           ...  
119385      673             0  
119386     1578             0  
119387     1104             0  
119388      731             0  
119389     1361             0  

[119390 rows x 17 columns]

In [196]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  object        
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  object        
 8   Customer Type         119390 non-null  object        
 9   Country               118902 non-null  object        
 10  Deposit Type          119390 non-null  object        
 11  Avg Daily Rate        119390 non-null  int64         
 12  Status                119390 non-null  object        
 13 

After renaming the column, it has not changed to the new name. To do that, check next cell

In [197]:
SHG_Booking.rename(columns={'Cancelled (0/1)':'Cancelled'}, inplace=True)

In [198]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  object        
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  object        
 8   Customer Type         119390 non-null  object        
 9   Country               118902 non-null  object        
 10  Deposit Type          119390 non-null  object        
 11  Avg Daily Rate        119390 non-null  int64         
 12  Status                119390 non-null  object        
 13 

we can clearly see the result of adding "inplace=True". Next, we change the data type

In [199]:
SHG_Booking['Cancelled']=SHG_Booking['Cancelled'].astype({'Cancelled':bool})

In [200]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  object        
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  object        
 8   Customer Type         119390 non-null  object        
 9   Country               118902 non-null  object        
 10  Deposit Type          119390 non-null  object        
 11  Avg Daily Rate        119390 non-null  int64         
 12  Status                119390 non-null  object        
 13 

In [201]:
SHG_Booking

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
0                1  Resort   2014-07-24   2015-07-01        342       0   
1                2  Resort   2013-06-24   2015-07-01        737       0   
2                3  Resort   2015-06-24   2015-07-01          7       1   
3                4  Resort   2015-06-18   2015-07-01         13       1   
4                5  Resort   2015-06-17   2015-07-01         14       2   
...            ...     ...          ...          ...        ...     ...   
119385      119386    City   2017-08-07   2017-08-30         23       7   
119386      119387    City   2017-05-21   2017-08-31        102       7   
119387      119388    City   2017-07-28   2017-08-31         34       7   
119388      119389    City   2017-05-14   2017-08-31        109       7   
119389      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            2                Direct     Transient        Portugal   
1            2                Direct     Transient        Portugal   
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update  Cancelled  \
0        No Deposit               0  Check-Out    2015-07-01      False   
1        No Deposit               0  Check-Out    2015-07-01      False   
2        No Deposit              75  Check-Out    2015-07-02      False   
3        No Deposit              75  Check-Out    2015-07-02      False   
4        No Deposit              98  Check-Out    2015-07-03      False   
...             ...             ...        ...           ...        ...   
119385   No Deposit              96  Check-Out    2017-09-06      False   
119386   No Deposit             225  Check-Out    2017-09-07      False   
119387   No Deposit             158  Check-Out    2017-09-07      False   
119388   No Deposit             104  Check-Out    2017-09-07      False   
119389   No Deposit             151  Check-Out    2017-09-07      False   

        Revenue  Revenue Loss  
0             0             0  
1             0             0  
2            75             0  
3            75             0  
4           196             0  
...         ...           ...  
119385      673             0  
119386     1578             0  
119387     1104             0  
119388      731             0  
119389     1361             0  

[119390 rows x 17 columns]

I just wish for more relatablity, so I want to change the input data of the column "Cancelled" to be either Yes/No  but this will mean that the data type will change from "bool" to "object".

In [202]:
SHG_Booking['Cancelled']=SHG_Booking['Cancelled'].map({True:'Yes', False:'No'})

In [203]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  object        
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  object        
 8   Customer Type         119390 non-null  object        
 9   Country               118902 non-null  object        
 10  Deposit Type          119390 non-null  object        
 11  Avg Daily Rate        119390 non-null  int64         
 12  Status                119390 non-null  object        
 13 

In [204]:
SHG_Booking

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
0                1  Resort   2014-07-24   2015-07-01        342       0   
1                2  Resort   2013-06-24   2015-07-01        737       0   
2                3  Resort   2015-06-24   2015-07-01          7       1   
3                4  Resort   2015-06-18   2015-07-01         13       1   
4                5  Resort   2015-06-17   2015-07-01         14       2   
...            ...     ...          ...          ...        ...     ...   
119385      119386    City   2017-08-07   2017-08-30         23       7   
119386      119387    City   2017-05-21   2017-08-31        102       7   
119387      119388    City   2017-07-28   2017-08-31         34       7   
119388      119389    City   2017-05-14   2017-08-31        109       7   
119389      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            2                Direct     Transient        Portugal   
1            2                Direct     Transient        Portugal   
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update Cancelled  \
0        No Deposit               0  Check-Out    2015-07-01        No   
1        No Deposit               0  Check-Out    2015-07-01        No   
2        No Deposit              75  Check-Out    2015-07-02        No   
3        No Deposit              75  Check-Out    2015-07-02        No   
4        No Deposit              98  Check-Out    2015-07-03        No   
...             ...             ...        ...           ...       ...   
119385   No Deposit              96  Check-Out    2017-09-06        No   
119386   No Deposit             225  Check-Out    2017-09-07        No   
119387   No Deposit             158  Check-Out    2017-09-07        No   
119388   No Deposit             104  Check-Out    2017-09-07        No   
119389   No Deposit             151  Check-Out    2017-09-07        No   

        Revenue  Revenue Loss  
0             0             0  
1             0             0  
2            75             0  
3            75             0  
4           196             0  
...         ...           ...  
119385      673             0  
119386     1578             0  
119387     1104             0  
119388      731             0  
119389     1361             0  

[119390 rows x 17 columns]

In [205]:
SHG_Booking.describe(include='all')

Booking ID   Hotel                   Booking Date  \
count   119390.000000  119390                         119390   
unique            NaN       2                            NaN   
top               NaN    City                            NaN   
freq              NaN   79330                            NaN   
mean     59695.500000     NaN  2016-05-16 16:23:19.353379840   
min          1.000000     NaN            2013-06-24 00:00:00   
25%      29848.250000     NaN            2015-11-28 00:00:00   
50%      59695.500000     NaN            2016-05-04 00:00:00   
75%      89542.750000     NaN            2016-12-09 00:00:00   
max     119390.000000     NaN            2017-08-31 00:00:00   
std      34465.068657     NaN                            NaN   

                         Arrival Date      Lead Time         Nights  \
count                          119390  119390.000000  119390.000000   
unique                            NaN            NaN            NaN   
top                               NaN            NaN            NaN   
freq                              NaN            NaN            NaN   
mean    2016-08-28 16:39:45.727447808     104.011416       3.427900   
min               2015-07-01 00:00:00       0.000000       0.000000   
25%               2016-03-13 00:00:00      18.000000       2.000000   
50%               2016-09-06 00:00:00      69.000000       3.000000   
75%               2017-03-18 00:00:00     160.000000       4.000000   
max               2017-08-31 00:00:00     737.000000      69.000000   
std                               NaN     106.863097       2.557439   

               Guests Distribution Channel Customer Type   Country  \
count   119390.000000               119390        119390    118902   
unique            NaN                    5             4       174   
top               NaN  Online Travel Agent     Transient  Portugal   
freq              NaN                74072         89613     48590   
mean         1.968239                  NaN           NaN       NaN   
min          0.000000                  NaN           NaN       NaN   
25%          2.000000                  NaN           NaN       NaN   
50%          2.000000                  NaN           NaN       NaN   
75%          2.000000                  NaN           NaN       NaN   
max         55.000000                  NaN           NaN       NaN   
std          0.722394                  NaN           NaN       NaN   

       Deposit Type  Avg Daily Rate     Status                  Status Update  \
count        119390   119390.000000     119390                         119390   
unique            3             NaN          3                            NaN   
top      No Deposit             NaN  Check-Out                            NaN   
freq         104641             NaN      75166                            NaN   
mean            NaN      101.826610        NaN  2016-07-30 00:24:47.883407104   
min             NaN       -6.000000        NaN            2014-10-17 00:00:00   
25%             NaN       69.000000        NaN            2016-02-01 00:00:00   
50%             NaN       95.000000        NaN            2016-08-07 00:00:00   
75%             NaN      126.000000        NaN            2017-02-08 00:00:00   
max             NaN     5400.000000        NaN            2017-09-14 00:00:00   
std             NaN       50.537273        NaN                            NaN   

       Cancelled        Revenue   Revenue Loss  
count     119390  119390.000000  119390.000000  
unique         2            NaN            NaN  
top           No            NaN            NaN  
freq       75166            NaN            NaN  
mean         NaN     247.932716    -109.917841  
min          NaN     -64.000000   -6300.000000  
25%          NaN       0.000000       0.000000  
50%          NaN     176.000000       0.000000  
75%          NaN     343.000000       0.000000  
max          NaN    7590.000000       0.000000  
std          NaN     300.923285     

From the information above, I could see clearly that the following columns; Hotel, Distribution Channel, Customer Type, Country, Deposit Type, Status and Cancelled all have unique values. For memory optimization, I will change the data type to "category"

In [206]:
SHG_Booking['Hotel']=SHG_Booking['Hotel'].astype('category')
SHG_Booking['Distribution Channel']=SHG_Booking['Distribution Channel'].astype('category')
SHG_Booking['Customer Type']=SHG_Booking['Customer Type'].astype('category')
SHG_Booking['Country']=SHG_Booking['Country'].astype('category')
SHG_Booking['Deposit Type']=SHG_Booking['Deposit Type'].astype('category')
SHG_Booking['Cancelled']=SHG_Booking['Cancelled'].astype('category')
SHG_Booking['Status']=SHG_Booking['Status'].astype('category')

In [207]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  category      
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  category      
 8   Customer Type         119390 non-null  category      
 9   Country               118902 non-null  category      
 10  Deposit Type          119390 non-null  category      
 11  Avg Daily Rate        119390 non-null  int64         
 12  Status                119390 non-null  category      
 13 

Notice how the memory usage has changed from 15.5+MB to 10MB

In [208]:
SHG_Booking['Avg Daily Rate']=SHG_Booking['Avg Daily Rate'].astype(float)
SHG_Booking['Revenue']=SHG_Booking['Revenue'].astype(float)
SHG_Booking['Revenue Loss']=SHG_Booking['Revenue Loss'].astype(float)

This is important because we are dealing with financial calculations in these columns

In [209]:
SHG_Booking.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 17 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   Booking ID            119390 non-null  int64         
 1   Hotel                 119390 non-null  category      
 2   Booking Date          119390 non-null  datetime64[ns]
 3   Arrival Date          119390 non-null  datetime64[ns]
 4   Lead Time             119390 non-null  int64         
 5   Nights                119390 non-null  int64         
 6   Guests                119390 non-null  int64         
 7   Distribution Channel  119390 non-null  category      
 8   Customer Type         119390 non-null  category      
 9   Country               118902 non-null  category      
 10  Deposit Type          119390 non-null  category      
 11  Avg Daily Rate        119390 non-null  float64       
 12  Status                119390 non-null  category      
 13 

To ensure/increase data quality we must be sure that the Booking ID has no duplicates. 

In [210]:
SHG_Booking['Booking ID'].duplicated().sum()

0

This shows uniqueness. Let me check the uniqueness of values of each categorical data type. I want to know maybe another unique value is due to lower case, space and so on.

In [211]:
Column=['Hotel','Distribution Channel', 'Customer Type','Country', 'Deposit Type','Status','Cancelled']

for i in Column:
    print(f'The unique values in {i} are:')
    print(SHG_Booking[i].unique())
    print('='*50)

The unique values in Hotel are:
['Resort', 'City']
Categories (2, object): ['City', 'Resort']
The unique values in Distribution Channel are:
['Direct', 'Corporate', 'Online Travel Agent', 'Offline Travel Agent', 'Undefined']
Categories (5, object): ['Corporate', 'Direct', 'Offline Travel Agent', 'Online Travel Agent', 'Undefined']
The unique values in Customer Type are:
['Transient', 'Contract', 'Transient-Party', 'Group']
Categories (4, object): ['Contract', 'Group', 'Transient', 'Transient-Party']
The unique values in Country are:
['Portugal', 'United Kingdom', 'United States', 'Spain', 'Ireland', ..., 'New Caledonia', 'Kiribati', 'Sudan', 'French Southern Territories', 'Sierra Leone']
Length: 175
Categories (174, object): ['Albania', 'Algeria', 'American Samoa', 'Andorra', ..., 'Viet Nam', 'Virgin Islands', 'Zambia', 'Zimbabwe']
The unique values in Deposit Type are:
['No Deposit', 'Refundable', 'Non Refundable']
Categories (3, object): ['No Deposit', 'Non Refundable', 'Refundable']

As I saw the spelling of "Vietnam" as "Viet Nam", I decided to check if there is no Vietnam in this data. Because of the number of assumed unique values Country has, I will only check for Country.

In [212]:
from difflib import SequenceMatcher

categories = SHG_Booking['Country'].unique()
for i, cat1 in enumerate(categories):
    for cat2 in categories[i+1:]:
        ratio = SequenceMatcher(None, str(cat1).replace(' ', '').lower(), str(cat2).replace(' ', '').lower()).ratio()
        if ratio > 0.8 and cat1 != cat2:
            print(f"Potential duplicates: '{cat1}' and '{cat2}' (similarity: {ratio:.2f})")


Potential duplicates: 'Ireland' and 'Iceland' (similarity: 0.86)
Potential duplicates: 'Australia' and 'Austria' (similarity: 0.88)


With the result, we can be sure that there is no duplicate. And due to uniqueness confirmed, we can be sure that there are no invalid inputs in the following columns;'Hotel','Distribution Channel', 'Customer Type','Country', 'Deposit Type','Status', and 'Cancelled'. Let us check for missing values also


In [213]:
SHG_Booking.isna().sum()

Booking ID                0
Hotel                     0
Booking Date              0
Arrival Date              0
Lead Time                 0
Nights                    0
Guests                    0
Distribution Channel      0
Customer Type             0
Country                 488
Deposit Type              0
Avg Daily Rate            0
Status                    0
Status Update             0
Cancelled                 0
Revenue                   0
Revenue Loss              0
dtype: int64

Clearly, the column "Country" has 488 missing values. Now I am stuck on whether to drop the missing values or fill the values with "Unknown". I could have dropped the Country column now but because I want to have two cleaned data. One for my geo-analysis.

In [214]:
(SHG_Booking['Arrival Date']<SHG_Booking['Booking Date']).sum()

0

It is possible for the guest to arrive on the same day he/she books but not logical for the arrival date to be less than the booking date. Since it is zero, no need to drop.

In [215]:
(SHG_Booking['Lead Time'] !=(SHG_Booking['Arrival Date']-SHG_Booking['Booking Date']).dt.days).sum()

0

In [216]:
(SHG_Booking['Lead Time']>730).sum()

1

In [217]:
(SHG_Booking['Lead Time']>365).sum()

3148

We also confirm that Lead Time=Arrival Date-Booking Date for all data entries in the Lead Time column. The number of Lead Time above two years is very rare and above a year is also reasonable out of about 119390 data entries.

In [218]:
(SHG_Booking['Guests']<0).sum()

0

In [219]:
(SHG_Booking['Guests']==0).sum()

180

Though, the values of Guests is not negative but the fact that Guest==0 implies invalid booking . So we drop the rows

In [220]:
SHG_Booking=SHG_Booking[SHG_Booking['Guests']>0]

In [221]:
(SHG_Booking['Guests']==0).sum()

0

In [222]:
(SHG_Booking['Nights']<0).sum()

0

In [223]:
(SHG_Booking['Nights']==0).sum()

645

This increased the likelyhood of nights column having valid data because it is possible I book a Hotel for day use only but to be sure, I need to probe further.

In [224]:
Daily_use=SHG_Booking[(SHG_Booking['Nights']==0) &
            (SHG_Booking['Status']=='Check-Out') &
            (SHG_Booking['Arrival Date']==SHG_Booking['Status Update']) &
            (SHG_Booking['Avg Daily Rate']==SHG_Booking['Revenue']) &
             (SHG_Booking['Avg Daily Rate']>0) 
]

Daily_use.shape

(0, 17)

From this information, we can see that there is no entry which that points to daily use. This means that Nights must be greater than or equal to 1 

In [225]:
SHG_Booking=SHG_Booking[(SHG_Booking['Nights'])>=1]

In [226]:
(SHG_Booking['Nights']==0).sum()

0

I have dropped rows less than 1 Nights

In [227]:
SHG_Booking

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
2                3  Resort   2015-06-24   2015-07-01          7       1   
3                4  Resort   2015-06-18   2015-07-01         13       1   
4                5  Resort   2015-06-17   2015-07-01         14       2   
5                6  Resort   2015-06-17   2015-07-01         14       2   
6                7  Resort   2015-07-01   2015-07-01          0       2   
...            ...     ...          ...          ...        ...     ...   
119385      119386    City   2017-08-07   2017-08-30         23       7   
119386      119387    City   2017-05-21   2017-08-31        102       7   
119387      119388    City   2017-07-28   2017-08-31         34       7   
119388      119389    City   2017-05-14   2017-08-31        109       7   
119389      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
2            1                Direct     Transient  United Kingdom   
3            1             Corporate     Transient  United Kingdom   
4            2   Online Travel Agent     Transient  United Kingdom   
5            2   Online Travel Agent     Transient  United Kingdom   
6            2                Direct     Transient        Portugal   
...        ...                   ...           ...             ...   
119385       2  Offline Travel Agent     Transient         Belgium   
119386       3   Online Travel Agent     Transient          France   
119387       2   Online Travel Agent     Transient         Germany   
119388       2   Online Travel Agent     Transient  United Kingdom   
119389       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update Cancelled  \
2        No Deposit            75.0  Check-Out    2015-07-02        No   
3        No Deposit            75.0  Check-Out    2015-07-02        No   
4        No Deposit            98.0  Check-Out    2015-07-03        No   
5        No Deposit            98.0  Check-Out    2015-07-03        No   
6        No Deposit           107.0  Check-Out    2015-07-03        No   
...             ...             ...        ...           ...       ...   
119385   No Deposit            96.0  Check-Out    2017-09-06        No   
119386   No Deposit           225.0  Check-Out    2017-09-07        No   
119387   No Deposit           158.0  Check-Out    2017-09-07        No   
119388   No Deposit           104.0  Check-Out    2017-09-07        No   
119389   No Deposit           151.0  Check-Out    2017-09-07        No   

        Revenue  Revenue Loss  
2          75.0           0.0  
3          75.0           0.0  
4         196.0           0.0  
5         196.0           0.0  
6         214.0           0.0  
...         ...           ...  
119385    673.0           0.0  
119386   1578.0           0.0  
119387   1104.0           0.0  
119388    731.0           0.0  
119389   1361.0           0.0  

[118565 rows x 17 columns]

In [228]:
SHG_Booking[(SHG_Booking['Revenue'] < 0)].shape


(1, 17)

In [229]:
SHG_Booking[(SHG_Booking['Avg Daily Rate'] <0)].shape


(1, 17)

This shows that ADR is negative which does not make sense. So ,we need to remove.

In [230]:
SHG_Booking=SHG_Booking[SHG_Booking['Avg Daily Rate']>0]

In [231]:
SHG_Booking[(SHG_Booking['Avg Daily Rate'] <0)].shape


(0, 17)

Since ADR and revenue cannot be negative, it can either be zero or positive. For it to be zero, it means either the person did not show up or cancelled the booking.

In [232]:
SHG_Booking[(SHG_Booking['Avg Daily Rate'] == 0) &
            ( SHG_Booking['Status']=='Check-Out')
].shape

(0, 17)

In [233]:
SHG_Booking[(SHG_Booking['Revenue'] == 0.0) &
            ( SHG_Booking['Status']=='Check-Out')
].shape

(0, 17)

The above two cells does not work with business logic, so we must remove

In [234]:
SHG_Booking=SHG_Booking[~((SHG_Booking['Avg Daily Rate'] == 0) & (SHG_Booking['Status'] == 'Check-Out'))]

In [235]:
SHG_Booking[(SHG_Booking['Avg Daily Rate'] == 0.0) &
            ( SHG_Booking['Status']=='Check-Out')
].shape

(0, 17)

In [236]:
SHG_Booking[(SHG_Booking['Revenue'] == 0.0) &
            ( SHG_Booking['Status']=='Check-Out')
].shape

(0, 17)

I observe that SHG_Booking[(SHG_Booking['Revenue'] == 0) & ( SHG_Booking['Status']=='Check-Out']) have the same column as SHG_Booking[(SHG_Booking['Avg Daily Rate'] == 0) & ( SHG_Booking['Status']=='Check-Out')]

In [237]:
SHG_Booking[SHG_Booking['Status'] == 'Canceled']['Cancelled'].value_counts()


Cancelled
Yes    42829
No         0
Name: count, dtype: int64

Here, we can confirm that when the status indicate "Canceled", the Cancelled column indicate yes. Which is consistent.

In [238]:
print(SHG_Booking[SHG_Booking['Status Update'] < SHG_Booking['Booking Date']])

Empty DataFrame
Columns: [Booking ID, Hotel, Booking Date, Arrival Date, Lead Time, Nights, Guests, Distribution Channel, Customer Type, Country, Deposit Type, Avg Daily Rate, Status, Status Update, Cancelled, Revenue, Revenue Loss]
Index: []


This shows that there is no error

In [239]:
SHG_Booking[(SHG_Booking['Revenue Loss'] >0)].shape


(0, 17)

This shows hat the values in Revenue Loss column are less than or equal zero. Which makes sense. Let us investigate further to be sure we have a good data

In [240]:
SHG_Booking[SHG_Booking['Status'] == 'Canceled']['Revenue Loss'].value_counts()


Revenue Loss
 0.0       14460
-248.0       429
-124.0       286
-120.0       274
-126.0       246
           ...  
-1615.0        1
-3090.0        1
-2300.0        1
-2899.0        1
-1478.0        1
Name: count, Length: 1862, dtype: int64

In [241]:
SHG_Booking[SHG_Booking['Status'] == 'No-Show']['Revenue Loss'].value_counts()


Revenue Loss
-75.0      44
-183.0     35
 0.0       34
-65.0      15
-89.0      14
           ..
-1994.0     1
-323.0      1
-355.0      1
-156.0      1
-459.0      1
Name: count, Length: 505, dtype: int64

In [242]:
SHG_Booking[SHG_Booking['Status'] == 'Check-Out']['Revenue Loss'].value_counts()


Revenue Loss
0.0    73387
Name: count, dtype: int64

The fact that there is no revenue loss for check-outs makes sense. And also for when bookings are cancelled or customers do not show up relate to revenue loss and it is logically valid

In [243]:
SHG_Booking.isna().sum()

Booking ID                0
Hotel                     0
Booking Date              0
Arrival Date              0
Lead Time                 0
Nights                    0
Guests                    0
Distribution Channel      0
Customer Type             0
Country                 467
Deposit Type              0
Avg Daily Rate            0
Status                    0
Status Update             0
Cancelled                 0
Revenue                   0
Revenue Loss              0
dtype: int64

I just realized that it is not necessary I drop the Country column because replacing the null values with "Unknown" will do just fine and needed I drop the column, I can just manipulate the data for my analysis.

In [244]:
SHG_Booking['Country']=SHG_Booking['Country'].cat.add_categories('Unknown')

SHG_Booking['Country']=SHG_Booking['Country'].fillna('Unknown')

In [245]:
SHG_Booking.isna().sum()

Booking ID              0
Hotel                   0
Booking Date            0
Arrival Date            0
Lead Time               0
Nights                  0
Guests                  0
Distribution Channel    0
Customer Type           0
Country                 0
Deposit Type            0
Avg Daily Rate          0
Status                  0
Status Update           0
Cancelled               0
Revenue                 0
Revenue Loss            0
dtype: int64

In [246]:
SHG_Booking.reset_index(drop=True,inplace=True)

In [247]:
SHG_Booking

Booking ID   Hotel Booking Date Arrival Date  Lead Time  Nights  \
0                3  Resort   2015-06-24   2015-07-01          7       1   
1                4  Resort   2015-06-18   2015-07-01         13       1   
2                5  Resort   2015-06-17   2015-07-01         14       2   
3                6  Resort   2015-06-17   2015-07-01         14       2   
4                7  Resort   2015-07-01   2015-07-01          0       2   
...            ...     ...          ...          ...        ...     ...   
117392      119386    City   2017-08-07   2017-08-30         23       7   
117393      119387    City   2017-05-21   2017-08-31        102       7   
117394      119388    City   2017-07-28   2017-08-31         34       7   
117395      119389    City   2017-05-14   2017-08-31        109       7   
117396      119390    City   2017-02-05   2017-08-29        205       9   

        Guests  Distribution Channel Customer Type         Country  \
0            1                Direct     Transient  United Kingdom   
1            1             Corporate     Transient  United Kingdom   
2            2   Online Travel Agent     Transient  United Kingdom   
3            2   Online Travel Agent     Transient  United Kingdom   
4            2                Direct     Transient        Portugal   
...        ...                   ...           ...             ...   
117392       2  Offline Travel Agent     Transient         Belgium   
117393       3   Online Travel Agent     Transient          France   
117394       2   Online Travel Agent     Transient         Germany   
117395       2   Online Travel Agent     Transient  United Kingdom   
117396       2   Online Travel Agent     Transient         Germany   

       Deposit Type  Avg Daily Rate     Status Status Update Cancelled  \
0        No Deposit            75.0  Check-Out    2015-07-02        No   
1        No Deposit            75.0  Check-Out    2015-07-02        No   
2        No Deposit            98.0  Check-Out    2015-07-03        No   
3        No Deposit            98.0  Check-Out    2015-07-03        No   
4        No Deposit           107.0  Check-Out    2015-07-03        No   
...             ...             ...        ...           ...       ...   
117392   No Deposit            96.0  Check-Out    2017-09-06        No   
117393   No Deposit           225.0  Check-Out    2017-09-07        No   
117394   No Deposit           158.0  Check-Out    2017-09-07        No   
117395   No Deposit           104.0  Check-Out    2017-09-07        No   
117396   No Deposit           151.0  Check-Out    2017-09-07        No   

        Revenue  Revenue Loss  
0          75.0           0.0  
1          75.0           0.0  
2         196.0           0.0  
3         196.0           0.0  
4         214.0           0.0  
...         ...           ...  
117392    673.0           0.0  
117393   1578.0           0.0  
117394   1104.0           0.0  
117395    731.0           0.0  
117396   1361.0           0.0  

[117397 rows x 17 columns]

In [248]:
SHG_Booking.to_csv('SHG_Data.csv')